idb/typeinf.py

```diff
@@ -0,0 +1,1713 @@
+            self.csize = v_uint32()
+        else:
+            self.csize = None
+
+        self.buf = v_bytes()
+
+    def pcb_size(self):
+        self["buf"].vsSetLength(self.size)
+
+    def pcb_csize(self):
+        if self.csize is not None:
+            self["buf"].vsSetLength(self.csize)
+
+    def pcb_buf(self):
+        if self.csize is not None:
+            buf = zlib.decompress(self.buf)
+            self.vsSetField("buf", buf)
+        else:
+            buf = self.buf.tobytes() if isinstance(self.buf, memoryview) else self.buf
+
+        defs = []
+        offset = 0
+        for _ in range(self.ndefs):
+            _def = TILTypeInfo()
+            offset = _def.vsParse(buf, offset=offset)
+            defs.append(_def)
+        self.defs = defs
+
+    def find_by_name(self, name):
+        if not self.defs:
+            return None
+        _def = list(filter(lambda x: x.name == name, self.defs))
+        if len(_def) == 0:
+            return None
+        return _def[0]
+
+    def get_by_ordinal(self, ordinal):
+        return self.defs[ordinal - 1]
+
+
+TIL_ZIP = 0x0001  # pack buckets using zip
+TIL_MAC = 0x0002  # til has macro table
+TIL_ESI = 0x0004  # extended sizeof info (short, long, longlong)
+TIL_UNI = 0x0008  # universal til for any compiler
+TIL_ORD = 0x0010  # type ordinal numbers are present
+TIL_ALI = 0x0020  # type aliases are present (this bit is used only on the disk)
+TIL_MOD = 0x0040  # til has been modified, should be saved
+TIL_STM = 0x0080  # til has extra streams
+TIL_SLD = 0x0100  # sizeof(long double)
+
+
+class TIL(VStruct):
+    def __init__(self, buf=None, wordsize=4):
+        VStruct.__init__(self)
+        self.wordsize = wordsize
+        self.signature = v_str(size=0x06)
+
+        # https://github.com/aerosoul94/tilutil/blob/master/distil.py#L545
+
+        self.format = v_uint32()
+        self.flags = v_uint32()
+
+        self.title_len = v_uint8()
+        self.title = v_str()
+
+        self.base_len = v_uint8()
+        self.base = v_str()
+
+        self.id = v_uint8()
+        self.cm = v_uint8()
+        self.size_i = v_uint8()
+        self.size_b = v_uint8()
+        self.size_e = v_uint8()
+        self.def_align = v_uint8()
+
+        # self.size_s  uint8
+        # self.size_l  uint8
+        # self.size_ll  uint8
+        # self.size_ldbl  uint8
+        # self.syms  TILBucket
+        # self.type_ordinal_numbers  uint32
+        # self.types  TILBucket
+        # self.macros  TILBucket
+
+    def pcb_flags(self):
+        if self.flags & TIL_ESI:
+            self.vsAddField("size_s", v_uint8())
+            self.vsAddField("size_l", v_uint8())
+            self.vsAddField("size_ll", v_uint8())
+
+        if self.flags & TIL_SLD:
+            self.vsAddField("size_ldbl", v_uint8())
+
+        self.vsAddField("syms", TILBucket(flags=self.flags))
+
+        if self.flags & TIL_ORD:
+            self.vsAddField("type_ordinal_numbers", v_uint32())
```

Comment:
I may just be nitpicky, but `type_ordinal_numbers` is actually a member of `til_bucket_t` (See `til_t` `syms`, `types`, and `macros`). It might need to be moved to `TILBucket`.

Reply:
When I moved it into the bucket, there was an exception, so it may not be, I'm not sure

Reply:
I see you solved it like this

```
        self._syms = TILBucket(self._header.flag & 0xCF, self._stream)
        self._types = TILBucket(self._header.flag, self._stream)
        self._macros = TILBucket(self._header.flag & 0xCF, self._stream)
```

Reply:
Yes, it appears only the type bucket can have that field.